### ISSUE TITLE
Unexpected results with BigInt.isValidInt

### ISSUE THREAD

**Issue by ktullavik** [dart-lang/sdk#36057]
This is with DDC, chrome, SDK 2.2.0, Windows 10

From bigint.dart:

```
  /**
   * Whether this big integer can be represented as an `int` without losing
   * precision.
   *
   * Warning: this function may give a different result on
   * dart2js, dev compiler, and the VM, due to the differences in
   * integer precision.
   */
  bool get isValidInt;
```

At first I thought it would check whether bigint is in the range
[MIN_SAFE_INT, MAX_SAFE_INT] as defined by js. 

Figured it checks whether the js number implementation (or the specs I guess)
happens to represent this integer as itself.
But then it further broke my expectations:

```
    BigInt b = BigInt.parse('119007199254740991');

    int c = 500;
    while (c-- > 0) {

        if (b.isValidInt) {
            int i = b.toInt();

            if ('$b' != '$i') {
                print("$b != $i");
            }
        }

        b += BigInt.one;
    }
```

output:

119007199254740992 != 119007199254741000
119007199254741008 != 119007199254741010
119007199254741024 != 119007199254741020
119007199254741056 != 119007199254741060
119007199254741072 != 119007199254741070
119007199254741088 != 119007199254741090
...

**Comment by lrhn** [comment]
tl;dr, the issue is the `toString` method of JavaScript numbers, not their value.

The issue here is that the JS `Number.prototype.toString` doesn't return a numeral of the exact value. Instead it returns the numeral with the most trailing zeros that is still not closer to any other value representable by a double.

Any of the double literals in the range `119007199254740984` to `119007199254741000` will evaluate to the double ...0992.
So, when choosing a string representation, the `toString` can pick any of those, and it picks the one with the most trailing zeros (that makes more sense when the zeros are after the decimal point, but they do it for integers too). So, comparing the `toString` of a JS number and a `BigInt` is not the same as comparing their numeric values, because the JS number's `toString` differs from its exact value.
